<div style="text-align: center;">
<h2>INFSCI 2595 Machine Learning - Spring 2019 </h2>
<h1 style="font-size: 250%;">Assignment #2</h1>
<h3>Total points: 100 </h3>
<h2> Due by April 1st </h2>
</div>

In [1]:
# Type in your information in the double quotes
firstName = "Shanim"
lastName = "Manzoor"
pittID = "shm150"

# <h3>  Problem #1. Linear Discriminant Analysis (LDA)and Quadratic Discriminant Analysis(QDA) </h3> 
 ### [30 points]
 
Do not write a code for this part

<h4> Problem #1-1 [10 points]</h4>  <br>
We have a classification problem with one feature $(x)$ and K classes to be classified. The prior probability of
class $k$ is $𝜋_{k} = 𝑃(𝑌 = 𝑘)$. Assume that the feature in class k has Gaussian distribution of
mean $μ_{k}$ and variance $σ^2 (𝒩(μ,𝜎^{2}))$.The variance is the same for all classes. 
Prove that the Bayes’ classifier (that chooses class k with largest $𝑃(𝑌 = 𝑘|𝑥))$ is equivalent to assigning an observation to the class for which the discriminant function $𝛿_{k}(x)$ is
maximized, where 
\begin{array} \\
𝛿_{k}(x) = x\frac{\mu _{k}}{\sigma ^{2}}- \frac{\mu_{2}^{k}}{2\sigma ^{2}}+ log(\pi _{k})
\end{array}
<br> What is the name of this classifier?

<img src="files/file1.jpeg">

<h4> Problem #1-2 [15 points]</h4>  <br>
Extend **Problem #1-1** to include **p** features. With features from each class drawn from a
Gaussian distribution with mean vector $μ_{k}$ and covariance matrix $Σ_{k}$ (which is now
different for each class). What is the discriminant function that maximizes **𝑃(𝑌 = 𝑘|𝑥)**. Is
the relationship with the feature vector **x** linear?<br> What is this classifier?

<img src="files/file22.jpeg">

<h4> Problem #1-3 [5 points]</h4>  <br>
- Explain Bias and variance trade-off between the two above classifiers.

In LDA(Linear Discriminant Analysis) we see that if there are p predictors, it still assumes that the covariance matrix is same for all. If the variance is actually uniform everywhere, this would be correct and computation needed is much lower. But if the assumption is wrong , the results would be very poor. Hence it is much LESS FLEXIBLE AND HAS LESS VARIANCE BUT HAS HIGH BIAS.
In QDA (Quadratic Discriminant Analysis) we have separate co-variance matrix for each class which implies more computation . This model is hence MORE FLEXIBLE WHICH IMPLIES HIGH VARIANCE BUT LESS BIAS.

#### <h3>  Problem #2. Regularization    </h3>
### [15 points]

<h4> Problem #2-1. [3 points]</h4> Answer the following questions 

- What is the main purpose of using regularization?


The main purpose of regularization is to reduce the complexity of a model. This in turn decreases overfitting. Regularization works by adding a penalty term to our objective function  to control the complexity. This decreases variance of the model .
Further, we see that when the dataset has a huge number of features, while the number of instances are low in comparison or insufficient, we see that high variance starts coming into play. This can be controlled by regularization. It therefore decreases variance and increases bias.

<h4> Problem #2-2.Logistic Regression with Ridge Regularization [12 points] </h4>  <br>

In this part, you should download and analyze the Wisconson **"breast_cancer"** dataset. <br>

- Fit logistic regression model using ridge regularization with different values of  C = 0.1, 1, 5, 10, 50, 100, and 1000 (Note that C is the LogisticRegression function argument). For each value, report the estimated coefficients for the fitted model (do not just print summary, make a table with feature names and estimated coefficients)

- What happens to the coefficients as you increase C?

- What happens to the flexibility of the model as you increase C?

In [2]:
#no split done as not requested
from sklearn.datasets import load_breast_cancer
from sklearn.linear_model import LogisticRegression
import pandas as pd
cancer1 = load_breast_cancer()
coef_list = []
C_grid = [0.1, 1, 5, 10, 50, 100, 1000]
for c in C_grid:
    logistic = LogisticRegression(C = c).fit(cancer1.data, cancer1.target)
    coef_list.append(logistic.coef_.tolist()[0])
coef_df = pd.DataFrame(coef_list, columns=cancer1.feature_names , index = C_grid )
coef_df

C:\Users\shani\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\shani\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\shani\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\shani\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\shani\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence t

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0.1        0.665411      0.087508        0.310579  -0.011101        -0.024660   
1.0        2.101333      0.122048       -0.057356  -0.003594        -0.153787   
5.0        3.806946      0.165450       -0.316113  -0.007404        -0.535809   
10.0       4.399252      0.196342       -0.376778  -0.011085        -0.649935   
50.0       4.608842      0.234913       -0.385851  -0.014415        -0.674125   
100.0      5.891077      0.189023       -0.518216  -0.017789        -1.246230   
1000.0     5.901814      0.140402       -0.430363  -0.024759        -3.670967   

        mean compactness  mean concavity  mean concave points  mean symmetry  \
0.1            -0.112482       -0.159466            -0.068620      -0.034316   
1.0            -0.399851       -0.643594            -0.340199      -0.225912   
5.0            -0.662065       -1.431558            -1.068888      -0.755004   
10.0           -0.792452       -1.701317            -1.279704      -0.928721   
50.0           -0.875968       -1.797327            -1.315552      -0.979580   
100.0          -0.870716       -2.309676            -2.392791      -1.644340   
1000.0          0.233504       -4.288088            -6.835734      -3.846296   

        mean fractal dimension           ...             worst radius  \
0.1                  -0.006372           ...                 0.679468   
1.0                  -0.026004           ...                 1.245314   
5.0                   0.004385           ...                 0.755611   
10.0                  0.001637           ...                 0.398140   
50.0                 -0.011748           ...                 0.125515   
100.0                 0.127627           ...                -0.099841   
1000.0                0.821566           ...                -0.643605   

        worst texture  worst perimeter  worst area  worst smoothness  \
0.1         -0.228427        -0.212196   -0.015818         -0.045413   
1.0         -0.346269        -0.125742   -0.023993         -0.284648   
5.0         -0.455093        -0.004469   -0.024508         -1.021818   
10.0        -0.515016         0.053002   -0.023936         -1.239660   
50.0        -0.570265         0.090930   -0.022488         -1.283905   
100.0       -0.529985         0.049821   -0.018259         -2.418720   
1000.0      -0.493568         0.005924   -0.008779         -7.541171   

        worst compactness  worst concavity  worst concave points  \
0.1             -0.346369        -0.433008             -0.132559   
1.0             -1.118689        -1.571928             -0.653832   
5.0             -1.440923        -2.814447             -2.022093   
10.0            -1.763487        -3.339636             -2.427975   
50.0            -2.100238        -3.661979             -2.513409   
100.0           -1.094326        -2.969409             -4.439373   
1000.0           3.422716        -3.328733            -12.994694   

        worst symmetry  worst fractal dimension  
0.1          -0.109339                -0.032552  
1.0          -0.690302                -0.112870  
5.0          -1.989563                -0.104313  
10.0         -2.459531                -0.133186  
50.0         -2.656133                -0.175657  
100.0        -3.791072                 0.060051  
1000.0       -6.759584                 1.063963  

[7 rows x 30 columns]

# Comment:
 -as you increase C: C is inverse of the lambda parameter( logistic regression) , so we see that lower the C more the regularization. But once we INCREASE C, we can see the penalty becomes low and coefficient values increases(absolute). We can see this by comparing first and larst row of the table printed above this comment.
        
-Flexibility:
As C increases , we see that regularization effect decreases and hence flexibility increases

### <h3>  Problem #3. Logistic Regression and Unbalanced Datasets  </h3> 
### [25 points]


We fit a logistic regression model to predict the probability that an individual will default on his/her credit card balance. We used the total balance (single feature) to fit the model and got the results shown in the table below.

<h4> Problem #3-1. [5 points]</h4> Prediciton with Logistic regression <br>

# |Table|Coefficient|Std.error|Z-statistic|P-Value|
|:--:|:-------------------------------:|
|Intercept|-10.6513|0.3612|-29.5|<0.0001|
|balance|0.0055|0.002|24.9|<0.0001|

(You can refer to page 14 in the our slides "Logistic Regression.pdf" if the above table does not show well.)

- What is the parametric model used in logistic regression?
- What is the class label of an individual with a balance equals to 15,000 dollar? What is the class label of an individual with balance equals to 800 dollar? (Write a python function which takes two inputs (feature, model_parameters) and returns the class labels for the data). default is class 1 and non-default is class 0.

# Logistic regression is a parametric model which uses a form of the linear regression .
# It uses the LOGIT function which is linear with X( X is one of the classes).

In [3]:
import numpy as np
import math as math
#print(math.exp(3))
# def credit(feature,model_parameters):
#     num=model_parameters[0]
#     for i in range(len(feature)):
#          num+=feature[i]*model_parameters[i+1 ]
        
   
#     print(num)
#     prob=(1/(math.exp(-num) + 1))
#     print(prob)
#     if prob>0.5 :
#         return 1
#     else:
#         return 0
# feature1=[800]
# feature2=[15000]
# model_p1=[-10.6513,0.055 ]
# print(model_paramaters1)
# #case where feature =800
# status1=credit(feature1,model_p1)
# print(status1)
# status2=credit(feature2,model_p1)
# print(status2)

    

In [4]:
def def123(feature, model_parameters):
    result = model_parameters[0]
    for i in range(len(feature)):
        #modelparams are the b0 and b1 values ,feature is the balance here
        result += feature[i]*model_parameters[i+1]
        print(result)
    result = (1/(math.exp(-result) + 1))
    print('Probability is ',result)
    
    if(result > 0.5):
        return 1
    else:
        return 0
        
     
        

model_param1 = [-10.6513, 0.0055]
feature = [15000]
result = def123(feature, model_param1)
print('Induvidual with balance 15000 dollars is', result, '\n')


feature = [800]
result = def123(feature, model_param1)
print('Induvidual with balance o 800 dollars is  ', result, '\n')

71.8487
Probability is  1.0
Induvidual with balance 15000 dollars is 1 

-6.251300000000001
Probability is  0.0019242363522599306
Induvidual with balance o 800 dollars is   0 



<h4> Problem #3-2. [10 points]</h4>  <br>

The coefficients of logistic regression are obtained by maximizing the likelihood function

\begin{array} \\
l(\beta) = \prod_{i:y_{i}=1} P(y_{i} = 1|x)\prod_{i{}':y_{{i}'}=0} (1-P(y_{{i}'} = 1|x))
\end{array}
Show that maximizing the
likelihood function is equivalent to minimizing the cost function $J(\beta)$, such that.
\begin{array} \\
J(\beta) = -\sum [y_{i} log(P(y_{i} = 1|x)) + (1- y_{i})log(1- P(y_{i} = 1|x))]
\end{array}


Here $n$ is the number training examples. Mention one possible method for obtaining the
optimal coefficients.

<img src="files/file32.jpeg">

<h4> Problem #3-3 [10 points]</h4> <br>
In a fraud detection system, a QDA classifier’s confusion matrix is found to be:

|        |Predicted Class - Not fraud| Predicted Class - fraud|
|:--:|:-------------------------------:|
|Actual class – Not fraud|1200|25|
|Actual class – fraud|30|7|

- Is dataset balanced? Why?
- Evaluate the overall error rate <br>
- Evaluate the precision and the recall <br>


<img src="files/file4.jpeg">

### <h3>  Problem #4. SVM, Decision Trees, MLP Classification   </h3> 
### [30 points]

In this problem, you will use different classification methods, SVM, Decision Tree and MLP; and find their accuracies using the test data. 
We will also use the Wisconson **"breast_cancer"** dataset.
In all of the following subparts, use random_state=0 when you split the dataset into train and test and **standardize** the data.

In [5]:
# write your code here
import numpy as np
import pandas as pd
from sklearn import datasets 
from sklearn.metrics import confusion_matrix 
from sklearn.model_selection import train_test_split 
from sklearn.datasets import load_breast_cancer
  
dataset = load_breast_cancer()

X_train, X_test, y_train, y_test= train_test_split(dataset.data, dataset.target, random_state= 0)

# standardize data
from sklearn.preprocessing import StandardScaler 
scaler = StandardScaler().fit(X_train)
X_train_transformed= scaler.transform(X_train)
X_test_transformed= scaler.transform(X_test)

<h4> Problem #4-1.  Classification with SVM [10 points]</h4><br>
- How does the Radial Basis Function Kernel in SVM measure the similarity between a test point and a training example?
- Fit an SVM classifier with a radial basis function kernel, with gamma =0.1, and regularization parameter C set to 10. Use the classifier to predict class labels for the test data. 
- Calculate the accuracy and confusion matrix on the test data.

In [6]:
from sklearn.svm import SVC
svm1=SVC(kernel='rbf', gamma=0.1, C=10).fit(X_train_transformed,y_train)
pred1=svm1.predict(X_test_transformed)
print(pred1)

[0 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 1 0 0 0 0 0 1 1 0 1 1 0 1 0 1 0 1 0 1 0 1
 0 1 0 0 1 0 1 1 0 1 1 1 0 0 0 0 1 1 1 1 1 1 0 0 0 1 1 0 1 0 0 0 1 0 0 1 0
 0 1 1 1 1 1 0 0 0 1 0 1 1 1 0 0 1 0 1 0 1 1 0 1 1 1 0 1 1 1 0 0 0 1 0 0 1
 0 0 1 1 1 1 1 1 1 1 1 0 1 0 1 1 1 1 1 0 1 1 1 1 1 1 1 0 1 1 1 0]


In [7]:
from sklearn.metrics import confusion_matrix,precision_score, recall_score,accuracy_score
confusion=confusion_matrix(y_test,pred1)
print(confusion)
print(precision_score(y_test,pred1))
print(recall_score(y_test,pred1))
print(accuracy_score(y_test,pred1))

[[52  1]
 [ 4 86]]
0.9885057471264368
0.9555555555555556
0.965034965034965


<h4> Problem #4-2.  Classification with Decisin Tree (DT) [10 points]</h4><br>
- In this part, use DT classification method on the training data. Set the maximum depth of the tree to five. Then use the classifier to predict class labels for the test data. Calculate the accuracy and confusion matrix on the test data.
- Use Adaboost to combine four decision trees each of max_depth of five. Use random_state=0 in adaboost. Find the test accuracy. 

In [8]:
from sklearn.tree import DecisionTreeClassifier
treeModel=DecisionTreeClassifier(max_depth=5)
treeModel.fit(X_train_transformed,y_train)
pred2=treeModel.predict(X_test_transformed)
print(accuracy_score(y_test,pred2))
confusion2=confusion_matrix(y_test,pred2)
print(confusion2)

0.9020979020979021
[[50  3]
 [11 79]]


In [12]:
from sklearn.ensemble import AdaBoostClassifier
Boostmodel = AdaBoostClassifier(base_estimator=DecisionTreeClassifier(max_depth = 5), n_estimators=4, random_state = 0).fit(X_train_transformed, y_train)
print('The test accuracy is ', Boostmodel.score(X_test_transformed, y_test))
pred3=Boostmodel.predict(X_test_transformed)
confusion3=confusion_matrix(y_test,pred3)
print(confusion3)


The test accuracy is  0.9370629370629371
[[51  2]
 [ 7 83]]


<h4> Problem #4-3 [5 points]</h4>

Follow steps to answer questions to build a neural network using MLPClassifier from sklearn.neural_network. 
- Build a model that has two hidden layers, the first layer has 10 neurons and second layer has 5 neurons. 
- Use 'relu' activation function, and set the regularization parameter alpha=0.5. 
- Set max_iter=1000; Set the random_state=0.
- Use stochastic gradient descent (sgd) to solve the optimization  problem
- Report accuracy, confusion matrix, precision, and recall 

In [13]:
from sklearn.neural_network import MLPClassifier
mlp1=MLPClassifier(solver='sgd', activation='relu', random_state=0,hidden_layer_sizes=[10,5], alpha=0.5,max_iter=1000)
mlp1.fit(X_train_transformed, y_train)
print('The test accuracy is ', mlp1.score(X_test_transformed, y_test))
pred4=mlp1.predict(X_test_transformed)
confusion4=confusion_matrix(y_test,pred4)
print(confusion4)
print(precision_score(y_test,pred4))
print(recall_score(y_test,pred4))

The test accuracy is  0.9790209790209791
[[52  1]
 [ 2 88]]
0.9887640449438202
0.9777777777777777


<h4> Problem #4-4 [5 points]</h4>


Use the same setting as problem 4-3, but instead of using two hidden layers, use three hidden layers with 10, 8, 5 hidden neurons respectively.
- Find accuracy, precision, recall and confusion matrix. 
- Comment on the result comparing 4-3 and 4-4. 

In [11]:
from sklearn.neural_network import MLPClassifier
mlp1=MLPClassifier(solver='sgd', activation='relu', random_state=0,hidden_layer_sizes=[10,8,5], alpha=0.5,max_iter=1000)
mlp1.fit(X_train_transformed, y_train)
print('The test accuracy is ', mlp1.score(X_test_transformed, y_test))
pred5=mlp1.predict(X_test_transformed)
confusion5=confusion_matrix(y_test,pred5)
print(confusion5)
print(precision_score(y_test,pred5))
print(recall_score(y_test,pred5))

The test accuracy is  0.951048951048951
[[49  4]
 [ 3 87]]
0.9560439560439561
0.9666666666666667


# Comment
increasing the number of hidden layers in a neural network doesn't always translate to higher accuracy, as it depends on the complexity of the problem. Increasing layers beyond a certain point will lead to overfitting to the training data. This is why we see lower accuracy in 4.4 compared to 4.3 due to adding an unnecessary hidden layer